# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [43]:

import pandas as pd

#url_1 = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
df_1 = pd.read_csv('dataset_lab_cleaned.csv')

url_2 = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv'
df_2 = pd.read_csv(url_2)

url_3 = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv'
df_3 = pd.read_csv(url_3)


In [44]:
df_1 = df_1.drop(columns=['Unnamed: 0'])
#df_2 = df_2.drop(columns=['Unnamed: 0'])
#df_3 = df_3.drop(columns=['Unnamed: 0'])

In [45]:
def clean_column_names(df):
    """
    Cleans and standardizes column names in a DataFrame.
    """
    df.columns = [col.lower().replace(' ', '_') for col in df.columns]
    # Rename 'st' to 'state' for consistency
    if 'st' in df.columns:
        df = df.rename(columns={'st': 'state'})
    return df

In [46]:
# Use function to standardize columns names

df_2 = clean_column_names(df_2)
df_3 = clean_column_names(df_3)

In [47]:
# Handle individual datasets invalid values and formats

# df_2

df_2['customer_lifetime_value'] = df_2['customer_lifetime_value'].str.replace('%','').astype(float)
df_2['number_of_open_complaints'] = df_2['number_of_open_complaints'].str.extract('(\d)').astype(float)

<>:6: SyntaxWarning: invalid escape sequence '\d'
<>:6: SyntaxWarning: invalid escape sequence '\d'
C:\Users\Pedro\AppData\Local\Temp\ipykernel_25852\2341468947.py:6: SyntaxWarning: invalid escape sequence '\d'
  df_2['number_of_open_complaints'] = df_2['number_of_open_complaints'].str.extract('(\d)').astype(float)


In [ ]:
#df_3 - scaling 'customer_lifetime_value'
df_3['customer_lifetime_value'] = df_3['customer_lifetime_value'] * 100
df_3.head()


In [49]:
# Combine dataset:
#   Ensure column are consistent

df_2_reordered = df_2[df_1.columns]
df_3_reordered = df_3[df_1.columns]

In [ ]:
# Concatenate the 3 dataframes

full_df = pd.concat([df_1, df_2_reordered, df_3_reordered], ignore_index=True)
full_df


In [57]:
# handle null values 

def handle_null_values(df):
    """
    Fills null values in a DataFrame using appropriate strategies.
    """
    # Fill categorical columns with the mode
    for col in ['education', 'gender', 'policy_type', 'vehicle_class']:
        if df[col].isnull().any():
            mode_value = df[col].mode()[0]
            df[col] = df[col].fillna(mode_value)
            
    # Fill numerical columns with the median
    for col in ['customer_lifetime_value', 'income', 'monthly_premium_auto', 'total_claim_amount']:
        if df[col].isnull().any():
            median_value = df[col].median()
            df[col] = df[col].fillna(median_value)
            
    return df

In [ ]:
full_df = handle_null_values(full_df)
full_df.isnull()

In [ ]:
# Using .value_counts() method, one can assess that there are
# not duplicate values in the dataframe.

full_df.value_counts()

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [60]:
url_mkt = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv'
df_mkt = pd.read_csv(url_mkt)

In [76]:
df_mkt.to_csv('marketing_dataset.csv')

In [61]:
%load_ext autoreload
%autoreload 2

In [75]:
df_mkt.duplicated().sum()

np.int64(0)

In [81]:
df_mkt = df_mkt.drop(columns=['unnamed:_0'], axis = 1)

In [83]:
policies = {'Personal L1':'Personal', 'Personal L2':'Personal', 'Personal L3':'Personal',
             'Corporate L1':'Corporate', 'Corporate L2':'Corporate', 'Corporate L3':'Corporate',
             'Special L1':'Special', 'Special L2':'Special', 'Special L3':'Special'}
df_mkt['policy'] = df_mkt['policy'].replace(policies)

In [90]:
df_mkt

customer       state  customer_lifetime_value response  coverage  \
0      DK49336     Arizona              4809.216960       No     Basic   
1      KX64629  California              2228.525238       No     Basic   
2      LZ68649  Washington             14947.917300       No     Basic   
3      XL78013      Oregon             22332.439460      Yes  Extended   
4      QA50777      Oregon              9025.067525       No   Premium   
...        ...         ...                      ...      ...       ...   
10905  FE99816      Nevada             15563.369440       No   Premium   
10906  KX53892      Oregon              5259.444853       No     Basic   
10907  TL39050     Arizona             23893.304100       No  Extended   
10908  WA60547  California             11971.977650       No   Premium   
10909  IV32877  California              6857.519928       No     Basic   

      education effective_to_date employmentstatus gender  income  ...  \
0       College        2011-02-18         Employed      M   48029  ...   
1       College        2011-01-18       Unemployed      F       0  ...   
2      Bachelor        2011-02-10         Employed      M   22139  ...   
3       College        2011-01-11         Employed      M   49078  ...   
4      Bachelor        2011-01-17    Medical Leave      F   23675  ...   
...         ...               ...              ...    ...     ...  ...   
10905  Bachelor        2011-01-19       Unemployed      F       0  ...   
10906   College        2011-01-06         Employed      F   61146  ...   
10907  Bachelor        2011-02-06         Employed      F   39837  ...   
10908   College        2011-02-13         Employed      F   64195  ...   
10909  Bachelor        2011-01-08       Unemployed      M       0  ...   

      number_of_policies     policy_type     policy  renew_offer_type  \
0                      9  Corporate Auto  Corporate            Offer3   
1                      1   Personal Auto   Personal            Offer4   
2                      2   Personal Auto   Personal            Offer3   
3                      2  Corporate Auto  Corporate            Offer2   
4                      7   Personal Auto   Personal            Offer1   
...                  ...             ...        ...               ...   
10905                  7   Personal Auto   Personal            Offer3   
10906                  6   Personal Auto   Personal            Offer2   
10907                  2  Corporate Auto  Corporate            Offer1   
10908                  6   Personal Auto   Personal            Offer1   
10909                  3   Personal Auto   Personal            Offer4   

       sales_channel  total_claim_amount  vehicle_class vehicle_size  \
0              Agent          292.800000  Four-Door Car      Medsize   
1        Call Center          744.924331  Four-Door Car      Medsize   
2        Call Center          480.000000            SUV      Medsize   
3             Branch          484.013411  Four-Door Car      Medsize   
4             Branch          707.925645  Four-Door Car      Medsize   
...              ...                 ...            ...          ...   
10905            Web         1214.400000     Luxury Car      Medsize   
10906         Branch          273.018929  Four-Door Car      Medsize   
10907            Web          381.306996     Luxury SUV      Medsize   
10908         Branch          618.288849            SUV      Medsize   
10909            Web         1021.719397            SUV      Medsize   

      vehicle_type month  
0                A     2  
1                A     1  
2                A     2  
3                A     1  
4                A     1  
...            ...   ...  
10905            A     1  
10906            A     1  
10907            A     2  
10908            A     2  
10909            A     1  

[10910 rows x 26 columns]

In [91]:
df_mkt = df_mkt.astype({'months_since_last_claim': 'int64', 'months_since_policy_inception': 'int64', 'number_of_open_complaints':'int64'})

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [92]:
pivot_table = pd.pivot_table(
    df_mkt,
    values='total_claim_amount',
    index='sales_channel',
    aggfunc='sum'
)

In [94]:
pivot_table['total_claim_amount'] = pivot_table['total_claim_amount'].round(2)

In [96]:
highest_rev_channel = pivot_table['total_claim_amount'].max()
highest_rev = pivot_table['total_claim_amount'].max()

In [98]:
pivot_table

total_claim_amount
sales_channel                    
Agent                  1810226.82
Branch                 1301204.00
Call Center             926600.82
Web                     706600.04

- Top Performer: The Agent sales channel is the clear leader, generating the most total revenue with $1,810,226.82. This is significantly higher than any other channel.

- Second Place: The Branch sales channel comes in second with $1,301,204.00 in total revenue.

- Lowest Performers: The Call Center and Web channels generated the least revenue, with $926,600.82 and $706,600.04 respectively.

Key Insight: 
This analysis suggests that the Agent sales channel is the most effective in terms of generating revenue. This is a critical insight for the marketing department, as it highlights a channel that is performing well and should potentially be the focus of more investment and strategic planning. The lower performance of the Web and Call Center channels could indicate a need for a different strategy, such as optimizing their processes, improving the customer experience, or increasing marketing efforts through these channels.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [ ]:
pivot_table_clv = pd.pivot_table(
    df_mkt,
    values='customer_lifetime_value',
    index=['gender', 'education'],
    aggfunc='mean')

pivot_table_clv = pivot_table_clv.round(2)

In [ ]:
pivot_table_clv

In [103]:
top_group = pivot_table_clv['customer_lifetime_value'].idxmax()
top_clv = pivot_table_clv['customer_lifetime_value'].max()


In [104]:
lowest_group = pivot_table_clv['customer_lifetime_value'].idxmin()
lowest_clv = pivot_table_clv['customer_lifetime_value'].min()


In [105]:
print(f"\nAnalysis:")
print(f"The group with the highest average customer lifetime value is {top_group}, with an average of ${top_clv}.")
print(f"The group with the lowest average customer lifetime value is {lowest_group}, with an average of ${lowest_clv}.")
print("\nWe can also observe that within each education level, there isn't a massive difference between genders, indicating that education level might be a stronger factor in determining CLV.")



Analysis:
The group with the highest average customer lifetime value is ('F', 'High School or Below'), with an average of $8675.22.
The group with the lowest average customer lifetime value is ('F', 'Doctor'), with an average of $7328.51.

We can also observe that within each education level, there isn't a massive difference between genders, indicating that education level might be a stronger factor in determining CLV.


## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [107]:
# First: Change the column 'effective_to_date' to date time in order
# to use the data to find the month and year information

df_mkt['effective_to_date'] = pd.to_datetime(df_mkt['effective_to_date'])

df_mkt['month'] = df_mkt['effective_to_date'].dt.month


In [109]:
complaints_by_month_policy = df_mkt.groupby(['policy_type', 'month'])['number_of_open_complaints'].sum().reset_index()


complaints_by_month_policy.columns = ['policy_type', 'month', 'total_complaints']

print(complaints_by_month_policy)

      policy_type  month  total_complaints
0  Corporate Auto      1               415
1  Corporate Auto      2               361
2   Personal Auto      1              1635
3   Personal Auto      2              1363
4    Special Auto      1                84
5    Special Auto      2                91


From the table, the customer service department can easily see the number of complaints for each policy type, broken down by month. 
This is more useful than a total count, as it allows them to identify any seasonal trends or specific months where a certain policy type experiences a spike. 
    
    Example, if 'Personal Auto' complaints rise significantly in month 2 (February), the team could investigate if there were specific events or campaigns during that time that might have contributed to the increase.